blackjax/sgmcmc/gradients.py

```diff
@@ -53,9 +69,106 @@ def logposterior_estimator_fn(position: PyTree, data_batch: PyTree) -> float:
         logprior = logprior_fn(position)
         batch_loglikelihood = jax.vmap(loglikelihood_fn, in_axes=(None, 0))
         return logprior + data_size * jnp.mean(
-            batch_loglikelihood(position, data_batch), axis=0
+            batch_loglikelihood(position, minibatch), axis=0
+        )
+
+    def grad_estimator_fn(_, position, data_batch):
+        return jax.grad(logposterior_estimator_fn)(position, data_batch), None
+
+    return GradientEstimator(init_fn, grad_estimator_fn)
+
+
+class CVGradientState(NamedTuple):
+    """The state of the CV gradient estimator contains the gradient of the
+    Control Variate computed on the whole dataset at initialization.
+
+    """
+
+    control_variate_grad: PyTree
+
+
+def cv_grad_estimator(
+    logprior_fn: Callable,
+    loglikelihood_fn: Callable,
+    data: PyTree,
+    centering_position: PyTree,
+) -> GradientEstimator:
+    """Builds a control variate gradient estimator [1]_.
+
+    This algorithm was ported from [2]_.
+
+    Parameters
+    ----------
+    logprior_fn
+        The log-probability density function corresponding to the prior
+        distribution.
+    loglikelihood_fn
+        The log-probability density function corresponding to the likelihood.
+    data
+        The full dataset.
+    centering_position
+        Centering position for the control variates (typically the MAP).
+
+    References
+    ----------
+    .. [1]: Baker, J., Fearnhead, P., Fox, E. B., & Nemeth, C. (2019).
+            Control variates for stochastic gradient MCMC. Statistics
+            and Computing, 29(3), 599-615.
+    .. [2]: Coullon, J., & Nemeth, C. (2022). SGMCMCJax: a lightweight JAX
+            library for stochastic gradient Markov chain Monte Carlo algorithms.
+            Journal of Open Source Software, 7(72), 4113.
+
+    """
+    data_size = jax.tree_leaves(data)[0].shape[0]
+    logposterior_grad_estimator_fn = grad_estimator(
+        logprior_fn, loglikelihood_fn, data_size
+    ).estimate
```

Comment:
```suggestion
   simple_grad_estimator = grad_estimator(
        logprior_fn, loglikelihood_fn, data_size
    ).estimate
    logposterior_grad_estimator_fn = lambda *x: simple_grad_estimator(*x)[0]
```

Reply:
Not very happy of this suggestion as well so feel free to resolve.

Reply:
I agree that my code is not very satisfying too.